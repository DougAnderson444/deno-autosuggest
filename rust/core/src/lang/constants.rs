use super::CharClass;
use CharClass::{Consonant, Vowel};

pub const CHAR_CLASSES_LATIN: &[(CharClass, char)] = &[
    (Consonant, 'b'),
    (Consonant, 'c'),
    (Consonant, 'd'),
    (Consonant, 'f'),
    (Consonant, 'g'),
    (Consonant, 'h'),
    (Consonant, 'j'),
    (Consonant, 'k'),
    (Consonant, 'l'),
    (Consonant, 'm'),
    (Consonant, 'n'),
    (Consonant, 'p'),
    (Consonant, 'q'),
    (Consonant, 'r'),
    (Consonant, 's'),
    (Consonant, 't'),
    (Consonant, 'v'),
    (Consonant, 'w'),
    (Consonant, 'x'),
    (Consonant, 'z'),
    (Vowel, 'a'),
    (Vowel, 'e'),
    (Vowel, 'i'),
    (Vowel, 'o'),
    (Vowel, 'u'),
    (Vowel, 'y'),
];
