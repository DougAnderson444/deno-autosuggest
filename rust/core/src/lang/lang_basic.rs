#![allow(dead_code)]

use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_basic() -> Lang {
    let mut lang = Lang::new();
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }
    lang
}
