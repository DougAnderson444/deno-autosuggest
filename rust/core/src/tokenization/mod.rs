mod word;
mod text;

use crate::lang::{Lang, Chars, CharPattern};
pub use word::Word;
pub use text::Text;


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
