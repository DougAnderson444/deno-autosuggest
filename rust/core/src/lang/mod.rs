mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use rust_stemmers::Stemmer;
use crate::utils::to_vec;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 20;


pub struct Lang {
    stemmer:      Option<Stemmer>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
}


impl Lang {
    pub fn new() -> Self {
        Self {
            stemmer:      None,
            pos_map:      HashMap::default(),
            compose_map:  HashMap::default(),
            reduce_map:   HashMap::default(),
            stem_buffer:  RefCell::new(String::with_capacity(BUFFER_CAPACITY)),
            norm_buffer1: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
            norm_buffer2: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
        }
    }

    pub fn set_stemmer(&mut self, stemmer: Option<Stemmer>) {
        self.stemmer = stemmer;
    }

    pub fn add_pos(&mut self, word: &str, pos: PartOfSpeech) {
        let source   = to_vec(word);
        let composed = self.unicode_compose(&source).unwrap_or(source);
        let reduced  = self.unicode_reduce(&composed);

        self.pos_map.insert(composed, pos);

        if let Some((_, reduced)) = reduced {
            self.pos_map.insert(reduced, pos);
        }
    }

    pub fn add_unicode_composition(&mut self, from: &str, to: &str) {
        self.compose_map.insert(to_vec(from), to_vec(to));
    }

    pub fn add_unicode_reduction(&mut self, from: &str, to: &str) {
        self.reduce_map.insert(to_vec(from), to_vec(to));
    }

    pub fn stem(&self, word: &[char]) -> usize {
        if let Some(stemmer) = &self.stemmer {
            let buffer = &mut *self.stem_buffer.borrow_mut();
            buffer.clear();
            buffer.extend(word.iter());
            let stem = stemmer.stem(&buffer);
            stem.chars().count()
        } else {
            word.len()
        }
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn unicode_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn unicode_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::PartOfSpeech;
    use super::Lang;

    fn get_lang() -> Lang {
        let mut lang = Lang::new();
        lang.add_unicode_composition("ó", "ó");
        lang.add_unicode_reduction("ó", "o");
        lang.add_unicode_reduction("õ", "oo");
        lang.add_pos("fóo", PartOfSpeech::Particle);
        lang
    }

    #[test]
    fn unicode_compose_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_compose_nfd() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_reduced() {
        let input  = to_vec("foobar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_fill0() {
        let input  = to_vec("fõbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn get_pos_source() {
        let input = to_vec("fóo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_reduced() {
        let input = to_vec("foo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_unknown() {
        let input = to_vec("bar");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, None);
    }
}