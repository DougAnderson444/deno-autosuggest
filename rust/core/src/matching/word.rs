use crate::tokenization::{Word, WordView};
use super::WordMatch;
use super::damlev::DamerauLevenshtein;
use super::jaccard::Jaccard;


const LENGTH_THRESHOLD:  f64 = 0.26;
const JACCARD_THRESHOLD: f64 = 0.51;
const DAMLEV_THRESHOLD:  f64 = 0.21;

thread_local! {
    static DAMLEV:  DamerauLevenshtein = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>      = Jaccard::new();
}


pub fn word_match(rword: &WordView, qword: &WordView) -> Option<(WordMatch, WordMatch)> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword) {
        return None;
    }

    let mut best_match: Option<(WordMatch, WordMatch)> = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qword, rword);
        let dists = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                if qlen < qword.stem  { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);

                let rel = dist / max!(qlen, rlen, 1) as f64;
                if rel > DAMLEV_THRESHOLD { continue; }

                best_match = best_match
                    .take()
                    .filter(|m| m.0.typos <= dist)
                    .or_else(|| Some(WordMatch::new_pair(
                        rword,
                        qword,
                        rlen,
                        qlen,
                        dist,
                    )));

                if dist <= std::f64::EPSILON {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &WordView, qword: &WordView) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 1 || rlen <= 1 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &WordView, qword: &WordView) -> bool {
    let rslice = if qword.fin {
        rword.chars()
    } else {
        &rword.chars()[.. min!(qword.len() + 1, rword.len())]
    };
    let dist   = JACCARD.with(|j| j.rel_dist(rslice, qword.chars()));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::TextOwn;
    use crate::lang::{Lang, CharClass, lang_english};
    use super::{length_check, jaccard_check, word_match};


    fn text(s: &str) -> TextOwn {
        let lang = Lang::new();
        TextOwn::from_str(s).split(&CharClass::Whitespace, &lang)
    }

    #[test]
    fn length_check_len_4() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = length_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_len_7() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (false, "mai"),
            (false, "mail"),
            (false, "mailb"),
            (true,  "mailbo"),
            (true,  "mailbox"),
            (true,  "mailboxe"),
            (true,  "mailboxes"),
            (false, "mailboxess"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = length_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed length_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_4() {
        let sample = [
            (true,  "mail"),
            (true,  "bail"),
            (false, "bait"),
            (false, "balt"),
            (false, "bolt"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbot"),
            (true,  "railbot"),
            (false, "raidbot"),
            (false, "roidbot"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7_reduction() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbxx"),
            (true,  "mailxxx"),
            (true,  "maixxxx"),
            (false, "maxxxxx"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (true,  "mailbo"),
            (true,  "mailbox"),
            (true,  "mailboxe"),
            (false, "mailboxes"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = jaccard_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = length_check(&rtext.view(0), &qtext.view(0));
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qtext  = TextOwn::from_str("");
        let rtext  = TextOwn::from_str("");
        assert_eq!(word_match(&rtext.view(0), &qtext.view(0)), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qtext  = TextOwn::from_str("mailbox").fin(false);
        let rtext  = TextOwn::from_str("");
        assert_eq!(word_match(&rtext.view(0), &qtext.view(0)), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qtext  = TextOwn::from_str("").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_eq!(word_match(&rtext.view(0), &qtext.view(0)), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qtext  = TextOwn::from_str("mailbox").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qtext  = TextOwn::from_str("mailybox").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qtext  = TextOwn::from_str("mailox").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qtext  = TextOwn::from_str("maiblox").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qtext1 = TextOwn::from_str("mailboxes").fin(true);
        let qtext2 = TextOwn::from_str("mailboxes").fin(false);
        let rtext  = TextOwn::from_str("mail");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext1.view(0)));
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext2.view(0)));
    }

    #[test]
    fn match_word_full_stem() {
        let lang   = lang_english();
        let qtext1 = TextOwn::from_str("university").set_stem(&lang);
        let qtext2 = TextOwn::from_str("university");
        let rtext  = TextOwn::from_str("universe").set_stem(&lang);
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext1.view(0)));
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext2.view(0)));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qtext  = TextOwn::from_str("mailb").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qtext  = TextOwn::from_str("maiylb").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qtext  = TextOwn::from_str("maib").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qtext  = TextOwn::from_str("malib").fin(false);
        let rtext  = TextOwn::from_str("mailbox");
        assert_debug_snapshot!(word_match(&rtext.view(0), &qtext.view(0)));
    }
}
