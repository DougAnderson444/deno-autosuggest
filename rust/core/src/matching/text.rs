use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use crate::tokenization::{Word, TextRef};
use super::WordMatch;
use super::word::word_match;

thread_local! {
    static RMATCHES: RefCell<HashMap<usize, WordMatch>> = RefCell::new(HashMap::with_capacity_and_hasher(10, Default::default()));
    static QMATCHES: RefCell<HashMap<usize, WordMatch>> = RefCell::new(HashMap::with_capacity_and_hasher(10, Default::default()));
}


pub fn text_match(rtext: &TextRef, qtext: &TextRef) -> (Vec<WordMatch>, Vec<WordMatch>) {
    RMATCHES.with(|rcell| {
    QMATCHES.with(|qcell| {
        let rmatches = &mut *rcell.borrow_mut();
        let qmatches = &mut *qcell.borrow_mut();
        rmatches.clear();
        qmatches.clear();

        for qword in qtext.words.iter() {
            if qmatches.contains_key(&qword.ix) { continue; }
            let qword = qword.to_view(qtext);

            let mut candidate: Option<(WordMatch, WordMatch)> = None;

            for rword in rtext.words.iter() {
                if rmatches.contains_key(&rword.ix) { continue; }
                let rword = rword.to_view(rtext);
                let mut stop = false;

                None.or_else(|| {
                        let rnext = rtext.words.get(rword.ix + 1)?.to_view(rtext);
                        if qword.len() < rword.len() + rword.dist(&rnext) { return None; }
                        if rmatches.contains_key(&(rword.ix + 1)) { return None; }
                        let (rmatch,  qmatch)  = word_match(&rword.join(&rnext), &qword)?;
                        let (rmatch1, rmatch2) = rmatch.split(&rword, &rnext)?;
                        rmatches.insert(rmatch1.ix, rmatch1);
                        rmatches.insert(rmatch2.ix, rmatch2);
                        qmatches.insert(qmatch.ix,  qmatch);
                        candidate.take();
                        stop = true;
                        Some(())
                    })
                    .or_else(|| {
                        let qnext = qtext.words.get(qword.ix + 1)?.to_view(qtext);
                        if rword.len() < qword.len() + qword.dist(&qnext) { return None; }
                        if qmatches.contains_key(&(qword.ix + 1)) { return None; }
                        let (rmatch,  qmatch)  = word_match(&rword, &qword.join(&qnext))?;
                        let (qmatch1, qmatch2) = qmatch.split(&qword, &qnext)?;
                        rmatches.insert(rmatch.ix,  rmatch);
                        qmatches.insert(qmatch1.ix, qmatch1);
                        qmatches.insert(qmatch2.ix, qmatch2);
                        candidate.take();
                        stop = true;
                        Some(())
                    })
                    .or_else(|| {
                        let (rmatch, qmatch) = word_match(&rword, &qword)?;
                        if !candidate.is_some() || !rmatch.func {
                            stop = !rmatch.func;
                            candidate = Some((rmatch, qmatch));
                        }
                        Some(())
                    });

                if stop {
                    break;
                }
            }

            if let Some((rmatch, qmatch)) = candidate {
                rmatches.insert(rmatch.ix, rmatch);
                qmatches.insert(qmatch.ix, qmatch);
            }
        }

        let mut rmatches = rmatches.drain().map(|(_, m)| m).collect::<Vec<_>>();
        let mut qmatches = qmatches.drain().map(|(_, m)| m).collect::<Vec<_>>();
        rmatches.sort_by(|m1, m2| m1.ix.cmp(&m2.ix));
        qmatches.sort_by(|m1, m2| m1.ix.cmp(&m2.ix));

        (rmatches, qmatches)
    }) })
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::{Text, TextOwn};
    use crate::lang::{CharClass, lang_basic, lang_english};
    use super::{text_match};


    fn text(s: &str) -> TextOwn {
        let lang = lang_basic();
        Text::from_str(s)
            .split(&[CharClass::Punctuation, CharClass::Whitespace], &lang)
            .set_char_classes(&lang)
    }


    #[test]
    fn match_text_empty_both() {
        let rtext = text("");
        let qtext = text("").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let rtext = text("");
        let qtext = text("mailbox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let rtext = text("mailbox");
        let qtext = text("mailbox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let rtext = text("mailbox");
        let qtext = text("maiblox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let rtext = text("yellow mailbox");
        let qtext = text("yelow").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let rtext = text("yellow mailbox");
        let qtext = text("maiblox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let rtext = text("yellow mailbox");
        let qtext = text("maiblox yel").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let rtext = text("small yellow metal mailbox");
        let qtext = text("big malibox yelo").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut rtext = text("the theme");
        let mut qtext = text("the").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = lang_english();
        rtext = rtext.set_pos(&lang);
        qtext = qtext.set_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let rtext = text("sneaky");
        let qtext = text("sneak").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query() {
        let rtext = text("wifi router");
        let qtext = text("wi fi router").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_unfihished() {
        let rtext = text("microbiology");
        let qtext = text("micro bio").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_typos() {
        let rtext = text("microbiology");
        let qtext = text("mcro byology").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_short() {
        let rtext = text("t-light");
        let qtext = text("tli").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record() {
        let rtext = text("wi fi router");
        let qtext = text("wifi router").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_typos() {
        let rtext = text("micro biology");
        let qtext = text("mcrobiology").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_unfinished() {
        let rtext = text("micro biology");
        let qtext = text("microbio").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_regression_1() {
        let rtext = text("special, year");
        let qtext = text("especiall").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_regression_2() {
        let rtext1 = text("50's");
        let rtext2 = text("500w");
        let qtext  = text("50s").fin(false);
        assert_debug_snapshot!(text_match(&rtext1.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&rtext2.to_ref(), &qtext.to_ref()));
    }
}
