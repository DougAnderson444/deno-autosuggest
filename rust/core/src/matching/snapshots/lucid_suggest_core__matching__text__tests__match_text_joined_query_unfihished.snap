---
source: src/matching/text.rs
expression: "text_match(&rtext.to_ref(), &qtext.to_ref())"
---
(
    [
        WordMatch {
            ix: 0,
            len: 12,
            slice: (
                0,
                8,
            ),
            typos: 0.5,
            func: false,
            fin: false,
        },
    ],
    [
        WordMatch {
            ix: 0,
            len: 5,
            slice: (
                0,
                5,
            ),
            typos: 0.4,
            func: false,
            fin: true,
        },
        WordMatch {
            ix: 1,
            len: 3,
            slice: (
                0,
                3,
            ),
            typos: 0.1,
            func: false,
            fin: false,
        },
    ],
)
